services/background-worker/src/Wire/BackendNotificationPusher.hs

```diff
@@ -61,7 +72,21 @@ pushNotification targetDomain (msg, envelope) = do
           <> [ skipChanThreadKilled,
                logRetries (const $ pure True) logErrr
              ]
-  recovering policy handlers $ const go
+  -- Ensure that the mvars are reset correctly.
+  UnliftIO.bracket_ (takeMVar mvar) (putMVar mvar ()) $ do
+    -- Ensure that envelopes are acked if recovering still fails.
+    -- Otherwise Rabbit is going to think that we are still processing
+    -- it and won't send another message as we have set up the channel
+    -- to deliver a single message at a time and to wait for confirmation.
+    UnliftIO.onException
+      (recovering policy handlers (const go))
+      -- Reject for redelivery. This is needed for the following reasons
+      -- 1) We have a strict time limit when kubernetes sends SIGTERM. We need to stay within that.
+      -- 2) Just because we failed to push the notification _now_ doesn't mean we won't be able to
+      --    in a few minutes. Requeuing the message will give us another go at it, and the next pod
+      --    to gain single exclusive consumer will get this message.
+      $ lift
+      $ reject envelope True
```

Comment:
Messages are requeued in their original position, if possible. https://www.rabbitmq.com/nack.html

However I think we still want to use the MVars so that we can use the graceful shutdown time the kubernetes gives us to try and finish processing any given message before we gracefully shutdown the amqp connection. Without the cross-thread signalling we might tear down the network connection while a notification is being pushed out, before it can be acknowledged. Not having the MVars will increase re-delivery of messages when the pod is shutdown if anything has been delivered by http, but the connection is closed before amqp can signal that.

Reply:
I've added more comments to make this clearer. Without the MVars there is a brief window where messages could be delivered to endpoint services but not ACKed to Rabbit. This is a problem that we should try to avoid as much as possible.